automation/Extension/webext-instrumentation/src/background/http-instrument.ts

```diff
@@ -468,15 +468,47 @@ export class HttpInstrument {
       }
     }
     */
-    update.top_level_url = escapeUrl(tab.url);
+    update.top_level_url = this.getDocumentUrlForRequest(details);
     update.parent_frame_id = details.parentFrameId;
     update.frame_ancestors = escapeString(
       JSON.stringify(details.frameAncestors),
     );
-
     this.dataReceiver.saveRecord("http_requests", update);
   }
 
+  /**
+   * Code taken and adapted from
+   * https://github.com/EFForg/privacybadger/pull/2198/files
+   *
+   * Gets the URL for a given request's top-level document.
+   *
+   * The request's document may be different from the current top-level document
+   * loaded in tab as requests can come out of order:
+   *
+   * @param {Object} details chrome.webRequest request/response details object
```

Comment:
Nit: Instead of referring to chrome.webRequest we should refer to https://github.com/mozilla/OpenWPM/blob/master/automation/Extension/webext-instrumentation/src/types/browser-web-request-event-details.d.ts#L19, but instead of in tsdoc, it can be done as a typehint for the details parameter below as per https://github.com/mozilla/OpenWPM/blob/master/automation/Extension/webext-instrumentation/src/background/http-instrument.ts#L272